File: src/geometry_arc_length_sectors_medium.py
"""
geometry - arc_length_sectors (medium)
Generated: 2026-02-22T04:11:12.522768
"""

from problem_utils import *
from svg_utils import Diagram

ANGLE_POOL_DEG = [30, 45, 60, 72, 90, 108, 120, 135, 150, 180, 216, 240, 270]
RADIUS_POOL = list(range(2, 18))

def generate():
    problem_type = choice(['arc_length', 'sector_area', 'combined', 'find_radius', 'find_angle_deg'])

    if problem_type == 'arc_length':
        radius = choice(RADIUS_POOL)
        angle_deg = choice(ANGLE_POOL_DEG)
        arc_length = simplify(Rational(angle_deg, 360) * 2 * pi * radius)

        d = Diagram(width=280, height=280, padding=40)
        d.circle((0,0), radius)
        end_x = radius * cos(angle_deg * pi / 180)
        end_y = radius * sin(angle_deg * pi / 180)
        d.line((0,0),(radius,0))
        d.line((0,0),(end_x, end_y))
        d.arc((0,0), radius*0.3, 0, angle_deg)
        d.text(radius*0.4, radius*0.15, f"{angle_deg}°")
        d.segment_label((0,0),(radius,0),str(radius))

        return problem(
            question=f"Find the arc length of a circle with radius ${radius}$ and central angle ${angle_deg}°$. Express your answer in terms of $\\pi$.",
            answer=arc_length,
            difficulty=(1300, 1450),
            topic="geometry/arc_length_sectors",
            solution=steps(
                f"Arc length formula: $L = \\frac{{\\theta}}{{360°}} \\cdot 2\\pi r$",
                f"$L = \\frac{{{angle_deg}}}{{360}} \\cdot 2\\pi \\cdot {radius}$",
                f"$L = {latex(arc_length)}$"
            ),
            image=d.render()
        )

    elif problem_type == 'sector_area':
        radius = choice(RADIUS_POOL)
        angle_deg = choice(ANGLE_POOL_DEG)
        sector_area = simplify(Rational(angle_deg, 360) * pi * radius**2)

        d = Diagram(width=280, height=280, padding=40)
        d.circle((0,0), radius)
        end_x = radius * cos(angle_deg * pi / 180)
        end_y = radius * sin(angle_deg * pi / 180)
        d.line((0,0),(radius,0))
        d.line((0,0),(end_x, end_y))
        d.arc((0,0), radius, 0, angle_deg)
        d.segment_label((0,0),(radius*0.7, radius*0.4), str(radius))

        return problem(
            question=f"Find the area of a sector with radius ${radius}$ and central angle ${angle_deg}°$. Express your answer in terms of $\\pi$.",
            answer=sector_area,
            difficulty=(1300, 1450),
            topic="geometry/arc_length_sectors",
            solution=steps(
                f"Sector area formula: $A = \\frac{{\\theta}}{{360°}} \\cdot \\pi r^2$",
                f"$A = \\frac{{{angle_deg}}}{{360}} \\cdot \\pi \\cdot {radius}^2$",
                f"$A = \\frac{{{angle_deg}}}{{360}} \\cdot {radius**2}\\pi$",
                f"$A = {latex(sector_area)}$"
            ),
            image=d.render()
        )

    elif problem_type == 'combined':
        # Given arc length, find sector area
        radius = choice(RADIUS_POOL)
        angle_deg = choice([45, 60, 72, 90, 108, 120, 135, 150, 180])
        arc_length = simplify(Rational(angle_deg, 360) * 2 * pi * radius)
        sector_area = simplify(Rational(angle_deg, 360) * pi * radius**2)

        return problem(
            question=f"A sector of radius ${radius}$ has arc length ${latex(arc_length)}$. Find the area of the sector. Express your answer in terms of $\\pi$.",
            answer=sector_area,
            difficulty=(1450, 1600),
            topic="geometry/arc_length_sectors",
            solution=steps(
                f"Given $r = {radius}$ and $s = {latex(arc_length)}$",
                f"Find central angle: $\\theta = \\frac{{s}}{{r}} = \\frac{{{latex(arc_length)}}}{{{radius}}} = {latex(Rational(angle_deg,180)*pi)}$ rad",
                f"Convert to degrees: $\\theta = {angle_deg}°$",
                f"Sector area: $A = \\frac{{{angle_deg}}}{{360}} \\cdot \\pi \\cdot {radius}^2 = {latex(sector_area)}$"
            )
        )

    elif problem_type == 'find_radius':
        # Given arc length and angle, find radius
        angle_deg = choice(ANGLE_POOL_DEG)
        angle_rad = Rational(angle_deg, 180) * pi
        radius = choice([3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 15])
        arc_length = simplify(radius * angle_rad)

        return problem(
            question=f"An arc subtends a central angle of ${angle_deg}°$ and has length ${latex(arc_length)}$. Find the radius of the circle.",
            answer=radius,
            difficulty=(1400, 1550),
            topic="geometry/arc_length_sectors",
            solution=steps(
                f"Convert angle to radians: $\\theta = {angle_deg}° = {latex(angle_rad)}$ rad",
                f"Use $s = r\\theta$, so $r = \\frac{{s}}{{\\theta}}$",
                f"$r = \\frac{{{latex(arc_length)}}}{{{latex(angle_rad)}}} = {radius}$"
            ),
            calculator="scientific"
        )

    else:  # find_angle_deg
        # Given arc length and radius (integer answer in degrees)
        angle_deg = choice([30, 45, 60, 90, 120, 135, 150, 180])
        radius = choice(RADIUS_POOL)
        arc_length = simplify(Rational(angle_deg, 360) * 2 * pi * radius)

        return problem(
            question=f"A circle of radius ${radius}$ has an arc of length ${latex(arc_length)}$. Find the central angle in degrees.",
            answer=angle_deg,
            difficulty=(1350, 1500),
            topic="geometry/arc_length_sectors",
            solution=steps(
                f"Use $s = \\frac{{\\theta}}{{360°}} \\cdot 2\\pi r$",
                f"${latex(arc_length)} = \\frac{{\\theta}}{{360}} \\cdot 2\\pi \\cdot {radius}$",
                f"$\\theta = \\frac{{{latex(arc_length)} \\cdot 360}}{{2\\pi \\cdot {radius}}} = {angle_deg}°$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/geometry_arc_length_sectors_medium.jl
# geometry - arc_length_sectors (medium)
# Generated: 2026-03-08T20:07:02.317587

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script r θ begin
    set_topic!("geometry/arc_length_sectors")
    
    problem_type = rand(1:4)
    
    if problem_type == 1
        # Arc length given radius and central angle
        radius = randint(5, 25)
        angle_deg = choice([30, 45, 60, 90, 120, 135, 150, 180, 210, 240, 270, 300, 330])
        angle_rad = angle_deg * π / 180
        arc_length = radius * angle_rad
        
        d = DiagramObj(width=300, height=300)
        center_x, center_y = 150, 150
        circle!(d, center_x, center_y, radius * 2.5)
        point!(d, center_x, center_y, label="O", label_offset=(-15, -15))
        
        start_angle = 0
        end_angle = angle_deg
        line!(d, center_x, center_y, center_x + radius * 2.5, center_y, stroke="blue", stroke_width=2)
        end_x = center_x + radius * 2.5 * cosd(angle_deg)
        end_y = center_y - radius * 2.5 * sind(angle_deg)
        line!(d, center_x, center_y, end_x, end_y, stroke="blue", stroke_width=2)
        
        arc!(d, center_x, center_y, radius * 2.5, start_angle, end_angle, stroke="red", stroke_width=3)
        angle_arc!(d, center_x, center_y, 30, start_angle, end_angle, label="$(angle_deg)°")
        
        mid_angle = angle_deg / 2
        label_x = center_x + radius * 1.3 * cosd(mid_angle)
        label_y = center_y - radius * 1.3 * sind(mid_angle)
        text!(d, label_x, label_y, "r = $(radius)", font_size=14)
        
        problem(
            question="Find the arc length of a sector with radius $(radius) and central angle $(angle_deg)°. Express your answer in terms of π.",
            answer=arc_length,
            difficulty=(1200, 1400),
            solution=steps(
                "Convert angle to radians: \$\\theta = $(angle_deg)° \\cdot \\frac{\\pi}{180°} = $(tex(angle_rad))\$",
                sol("Arc length formula", "s = r\\theta"),
                sol("Substitute", "s = $(radius) \\cdot $(tex(angle_rad)) = $(tex(arc_length))")
            ),
            image=render(d),
            time=90
        )
        
    elseif problem_type == 2
        # Sector area given radius and central angle
        radius = randint(6, 20)
        angle_deg = choice([36, 45, 60, 72, 90, 120, 144, 150, 180])
        angle_rad = angle_deg * π / 180
        sector_area = radius^2 * angle_rad / 2
        
        d = DiagramObj(width=300, height=300)
        center_x, center_y = 150, 150
        circle!(d, center_x, center_y, radius * 2.8, stroke_dasharray="5,5")
        point!(d, center_x, center_y, label="O", label_offset=(-15, -15))
        
        start_angle = 0
        end_angle = angle_deg
        line!(d, center_x, center_y, center_x + radius * 2.8, center_y, stroke="blue", stroke_width=2)
        end_x = center_x + radius * 2.8 * cosd(angle_deg)
        end_y = center_y - radius * 2.8 * sind(angle_deg)
        line!(d, center_x, center_y, end_x, end_y, stroke="blue", stroke_width=2)
        
        arc!(d, center_x, center_y, radius * 2.8, start_angle, end_angle, stroke="blue", stroke_width=2)
        
        n_fill = 8
        for i in 0:(n_fill-1)
            fill_angle = start_angle + (end_angle - start_angle) * i / n_fill
            fill_x = center_x + radius * 1.4 * cosd(fill_angle)
            fill_y = center_y - radius * 1.4 * sind(fill_angle)
            point!(d, fill_x, fill_y, radius=2, color="lightblue")
        end
        
        angle_arc!(d, center_x, center_y, 35, start_angle, end_angle, label="$(angle_deg)°")
        text!(d, center_x + radius * 1.5, center_y + 30, "r = $(radius)", font_size=14)
        
        problem(
            question="Find the area of a sector with radius $(radius) and central angle $(angle_deg)°. Express your answer in terms of π.",
            answer=sector_area,
            difficulty=(1300, 1500),
            solution=steps(
                "Convert angle to radians: \$\\theta = $(angle_deg)° \\cdot \\frac{\\pi}{180°} = $(tex(angle_rad))\$",
                sol("Sector area formula", "A = \\frac{1}{2}r^2\\theta"),
                sol("Substitute", "A = \\frac{1}{2}($(radius))^2 \\cdot $(tex(angle_rad)) = $(tex(sector_area))")
            ),
            image=render(d),
            time=100
        )
        
    elseif problem_type == 3
        # Find central angle given arc length and radius
        radius = randint(8, 18)
        angle_deg = choice([30, 45, 60, 90, 120, 135, 150])
        angle_rad = angle_deg * π / 180
        arc_length = radius * angle_rad
        
        d = DiagramObj(width=300, height=300)
        center_x, center_y = 150, 150
        circle!(d, center_x, center_y, radius * 2.6, stroke_dasharray="5,5")
        point!(d, center_x, center_y, label="O", label_offset=(-15, -15))
        
        start_angle = 0
        end_angle = angle_deg
        line!(d, center_x, center_y, center_x + radius * 2.6, center_y, stroke="blue", stroke_width=2)
        end_x = center_x + radius * 2.6 * cosd(angle_deg)
        end_y = center_y - radius * 2.6 * sind(angle_deg)
        line!(d, center_x, center_y, end_x, end_y, stroke="blue", stroke_width=2)
        
        arc!(d, center_x, center_y, radius * 2.6, start_angle, end_angle, stroke="red", stroke_width=3)
        
        mid_angle = angle_deg / 2
        label_x = center_x + radius * 1.5 * cosd(mid_angle)
        label_y = center_y - radius * 1.5 * sind(mid_angle)
        text!(d, label_x, label_y, "s = $(tex(arc_length))", font_size=13)
        text!(d, center_x + radius * 1.3, center_y + 35, "r = $(radius)", font_size=14)
        
        problem(
            question="An arc has length \$$(tex(arc_length))\$ and the radius is $(radius). Find the central angle in degrees.",
            answer=angle_deg,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Arc length formula", "s = r\\theta"),
                sol("Solve for θ (in radians)", "\\theta = \\frac{s}{r} = \\frac{$(tex(arc_length))}{$(radius)} = $(tex(angle_rad))"),
                "Convert to degrees: \$\\theta = $(tex(angle_rad)) \\cdot \\frac{180°}{\\pi} = $(angle_deg)°\$"
            ),
            image=render(d),
            time=120
        )
        
    else
        # Find radius given sector area and central angle
        angle_deg = choice([40, 45, 60, 72, 90, 120, 135, 144])
        angle_rad = angle_deg * π / 180
        radius = randint(8, 16)
        sector_area = radius^2 * angle_rad / 2
        
        d = DiagramObj(width=300, height=300)
        center_x, center_y = 150, 150
        circle!(d, center_x, center_y, radius * 2.5, stroke_dasharray="5,5")
        point!(d, center_x, center_y, label="O", label_offset=(-15, -15))
        
        start_angle = 0
        end_angle = angle_deg
        line!(d, center_x, center_y, center_x + radius * 2.5, center_y, stroke="blue", stroke_width=2)
        end_x = center_x + radius * 2.5 * cosd(angle_deg)
        end_y = center_y - radius * 2.5 * sind(angle_deg)
        line!(d, center_x, center_y, end_x, end_y, stroke="blue", stroke_width=2)
        
        arc!(d, center_x, center_y, radius * 2.5, start_angle, end_angle, stroke="blue", stroke_width=2)
        
        angle_arc!(d, center_x, center_y, 35, start_angle, end_angle, label="$(angle_deg)°")
        text!(d, center_x - 20, center_y - 40, "A = $(tex(sector_area))", font_size=13)
        
        problem(
            question="A sector has area \$$(tex(sector_area))\$ and central angle $(angle_deg)°. Find the radius.",
            answer=radius,
            difficulty=(1500, 1700),
            solution=steps(
                "Convert angle to radians: \$\\theta = $(angle_deg)° \\cdot \\frac{\\pi}{180°} = $(tex(angle_rad))\$",
                sol("Sector area formula", "A = \\frac{1}{2}r^2\\theta"),
                sol("Solve for r", "r^2 = \\frac{2A}{\\theta} = \\frac{2 \\cdot $(tex(sector_area))}{$(tex(angle_rad))} = $(radius^2)"),
                sol("Answer", "r = $(radius)")
            ),
            image=render(d),
            time=130
        )
    end
end